Aggregation refactor: make aggregationFactory implement NamedWritable

colings86 opened the pull request:
Also makes AggregatorFactories implement Writable

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+            PipelineAggregatorFactory factory = in.readPipelineAggregatorFactory();
+            pipelineAggregatorFactoriesList.add(factory);
+        }
+        AggregatorFactories aggregatorFactories = new AggregatorFactories(
```
a bit weird to create an array list with a known size, just to convert it to an array. Let's build an array directly?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+        boolean hasParent = parent != null;
+        out.writeBoolean(hasParent);
+        if (hasParent) {
+            out.writeAggregatorFactory(parent);
```
I suspect there is an issue with this. Imagine we have a two-level aggregation, then the top-level aggs will be written twice, once on its own and once as a parent of its sub aggregations?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+                return false;
+        } else if (!pipelineAggregatorFactories.equals(other.pipelineAggregatorFactories))
+            return false;
+        return true;
```
we need to make these equals impls simpler. Maybe we can remove the "this == obj" optimization, use Objects.equal to avoid null checks, and remove the isAssignableFrom for classes that are never supposed to be equal to sub-classes (all?)?

colings86 replied:
so you mean something like this?

```java
if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        AggregatorFactories other = (AggregatorFactories) obj;
        if (!Objects.deepEquals(factories, other.factories)) 
            return false;
        if (!Objects.equals(pipelineAggregatorFactories, other.factories)) 
            return false;
        return true;
```

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+        result = prime * result + Arrays.hashCode(factories);
+        result = prime * result + ((parent == null) ? 0 : parent.hashCode());
+        result = prime * result + ((pipelineAggregatorFactories == null) ? 0 : pipelineAggregatorFactories.hashCode());
+        return result;
```
I think we can make it simpler using Objects.hash?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/support/ValueType.java:
```diff
@@ -19,26 +19,33 @@
+            IndexFieldData.class,
             ValueFormat.RAW),
-    LONG("byte|short|integer|long", ValuesSource.Numeric.class, IndexNumericFieldData.class, ValueFormat.RAW) {
+ LONG((byte) 2, "byte|short|integer|long", ValuesSource.Numeric.class,
```
indentation?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+            return false;
+        if (parent == null) {
+            if (other.parent != null)
+                return false;
```
I'm wondering if the parent really helps define equality here? Additionally, by adding this we will do more checks than necessary given that we compare both sub-aggs and parent aggs

colings86 replied:
My theory here was that you could have the same aggs structure (names, sub-aggs, options) in two different parts of the aggregation tree and they wouldn't be equal because they had different parents. So equality basically included the position in the aggs tree. But maybe thats not really right because then we are using an outside property (position in the aggs tree) to determine equality of an object?

jpountz commented:
I gave it a quick look and I think the fact that most aggregation objects track their parent is going to make the refactoring a bit less straightforward than for queries (see comments about serialization and checking for parents in equals)

colings86 commented:
@jpountz I've pushed a commit addressing your comments. If you could especially check the equals and hashCode methods in AggregatoryFactory, ValuesSourceAggregatorFactory, PipelineAggregatorFactory and AggregatorFactories, that would be great

javanna commented on core/src/test/java/org/elasticsearch/search/aggregations/BaseAggregationTestCase.java:
```diff
@@ -0,0 +1,259 @@
+        case 2:
+        default:
+            return INT_FIELD_NAME;
+        }
```
I wonder if we should start already sharing some common code between our BaseQueryTestCase and this class....wouldn't want to complicate things though. Also our base test class in not in master of course so that woul already complicate things...

javanna replied:
One more thing, maybe you want to have a testEqualsAndHashcode like we recently introduced in the query-refactoring branch, just to make sure that equals and hashcode are properly implemented.

javanna commented:
I had a look and this looks good to me, I think the main concern around equality checks has been addressed, all good besides that

jpountz commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -49,6 +51,9 @@
  */
+/**
+ *
+ */
```
duplicated empty javadocs? :)

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -251,4 +260,71 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            if (!(obj.getClass().isAssignableFrom(getClass()) || getClass().isAssignableFrom(obj.getClass())))
```
I don't think this second `if` statement is desirable/useful?

jpountz commented:
LGTM
